src/core/Akka.Streams/Implementation/Fusing/Ops.cs

```diff
@@ -3762,6 +3762,140 @@ public RecoverWith(Func<Exception, IGraph<SourceShape<TOut>, TMat>> partialFunct
     /// </summary>
     /// <typeparam name="TIn">TBD</typeparam>
     /// <typeparam name="TOut">TBD</typeparam>
+    [InternalApi]
+    public sealed class StatefulSelectManyAsync<TIn, TOut> : GraphStage<FlowShape<TIn, TOut>>
+    {
+        #region internal classes
+
+        private sealed class Logic : InAndOutGraphStageLogic
+        {
+            private readonly StatefulSelectManyAsync<TIn, TOut> _stage;
+            private IterableAdapter<TOut> _currentIterator;
+            private readonly Decider _decider;
+            private Func<TIn, IAsyncEnumerable<TOut>> _plainConcat;
+
+            public Logic(StatefulSelectManyAsync<TIn, TOut> stage, Attributes inheritedAttributes) : base(stage.Shape)
+            {
+                _stage = stage;
+                _decider = inheritedAttributes.GetAttribute(new ActorAttributes.SupervisionStrategy(Deciders.StoppingDecider)).Decider;
+                _plainConcat = stage._concatFactory();
+
+                SetHandler(stage._in, this);
+                SetHandler(stage._out, this);
+            }
+
+            public override void OnPush()
+            {
+                try
+                {
+                    _currentIterator = new IterableAdapter<TOut>(_plainConcat(Grab(_stage._in)).GetAsyncEnumerator());
+                    PushPull().Wait();
+                }
+                catch (Exception ex)
+                {
+                    var directive = _decider(ex);
+                    switch (directive)
+                    {
+                        case Directive.Stop:
+                            FailStage(ex);
+                            break;
+                        case Directive.Resume:
+                            if (!HasBeenPulled(_stage._in))
+                                Pull(_stage._in);
+                            break;
+                        case Directive.Restart:
+                            RestartState();
+                            if (!HasBeenPulled(_stage._in))
+                                Pull(_stage._in);
+                            break;
+                        default:
+                            throw new ArgumentOutOfRangeException();
+                    }
+                }
+            }
+
+            public override void OnUpstreamFinish()
+            {
+                if (!HasNext().Result)
```

Comment:
Again, don't block here. This can cause a deadlock when terminating the stream graph if the upstream `IAsyncEnumerable` ahs not been disposed. 

But the business logic is wrong anyway - there can't be an upstream if we're the `Source`.

Reply:
Actually, it looks like this stage is a Flow and not a Source - why is that?

Reply:
@Aaronontheweb  FWIW there are situations where it -may- make sense to unfold an async enumerator from an earlier stage... i.e. I wouldn't mind a `SelectMany` that takes `IAsyncEnumerable` from a previous stage and behaves the way the existing one does.

Agreed this probably needs a redesign though. I'd assume that handling an `IAsyncEnumerable` would be almost like a special case of `UnfoldResourceAsync`, at least in naieve implementation.